pyfr/integrators/dual/phys/steppers.py

```diff
@@ -1,25 +1,125 @@
 # -*- coding: utf-8 -*-
 
+from itertools import chain
+
 from pyfr.integrators.dual.phys.base import BaseDualIntegrator
 
 
 class BaseDualStepper(BaseDualIntegrator):
-    pass
 
+    def _set_currstg(self, s):
+        self.currstg = s
+
+    def _finalize_stage(self, tcurr):
+        pass
+
+
+class BaseBDFStepper(BaseDualStepper):
+    _nstages = 1
+
+    @property
+    def stage_nregs(self):
+        return 0
+
+    @property
+    def stepper_nregs(self):
+        return len(self.stepper_static_coeffs) - 1
+
+    def _finalize_stage(self, tcurr):
+        psnregs = self.pseudointegrator.pseudo_stepper_nregs
+        snregs = self.pseudointegrator.stepper_nregs
 
-class DualBDF2Stepper(BaseDualStepper):
+        # Rotate the source registers to the right by one
+        self.pseudointegrator._regidx[psnregs:psnregs + snregs] = (
+            self.pseudointegrator._stepper_regidx[-1:]
+            + self.pseudointegrator._stepper_regidx[:-1]
+        )
+
+        # Copy the current soln into the first source register
+        self.pseudointegrator._add(
+            0, self.pseudointegrator._stepper_regidx[0],
+            1, self.pseudointegrator._idxcurr
+        )
+
+    @property
+    def stepper_coeffs(self):
+        return [1] + [sc/self._dt for sc in self.stepper_static_coeffs]
+
+
+class DualBDF2Stepper(BaseBDFStepper):
     stepper_name = 'bdf2'
     stepper_order = 2
-    stepper_coeffs = [-1.5, 2.0, -0.5]
+    stepper_static_coeffs = [-1.5, 2.0, -0.5]
 
 
-class DualBDF3Stepper(BaseDualStepper):
+class DualBDF3Stepper(BaseBDFStepper):
     stepper_name = 'bdf3'
     stepper_order = 3
-    stepper_coeffs = [-11.0/6.0, 3.0, -1.5, 1.0/3.0]
+    stepper_static_coeffs = [-11.0/6.0, 3.0, -1.5, 1.0/3.0]
 
 
-class DualBackwardEulerStepper(BaseDualStepper):
+class DualBackwardEulerStepper(BaseBDFStepper):
     stepper_name = 'backward-euler'
     stepper_order = 1
-    stepper_coeffs = [-1.0, 1.0]
+    stepper_static_coeffs = [-1.0, 1.0]
+
+
+class BaseDIRKStepper(BaseDualStepper):
+    @property
+    def stepper_nregs(self):
+        return 1
+
+    @property
+    def stage_nregs(self):
+        return self._nstages
+
+    def _finalize_stage(self, tcurr):
+        # Finalize DIRK
+        if not self.fsal:
+            # Get the new soln
+            self.pseudointegrator._add(
+                0, self.pseudointegrator._idxcurr,
+                1, self.pseudointegrator._stepper_regidx[0],
+                *chain(*zip([bred*self._dt for bred in self.b],
+                            self.pseudointegrator._stage_regidx)),
+                subdims=self.pseudointegrator._subdims
+            )
+
+        # Copy the new solution into stepper register
+        self.pseudointegrator._add(
+            0, self.pseudointegrator._stepper_regidx[0],
+            1, self.pseudointegrator._idxcurr
+        )
+
+    @property
+    def stepper_coeffs(self):
```

Comment:
Is it possible to make stepper coeffs a function call with a stage parameter?

Reply:
This can be a good idea after we remove bdf steppers